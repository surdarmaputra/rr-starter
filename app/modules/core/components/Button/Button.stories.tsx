import type { Meta, StoryObj } from '@storybook/react';

import { Button, type ButtonProps } from './Button';

type Story = StoryObj<typeof Button>;

const meta: Meta = {
  component: Button,
  title: 'Core/Components/Button',
};

export default meta;

export const Default: Story = {
  args: {
    children: 'Button',
  },
  render: (args: ButtonProps) => <Button {...args} />,
};
