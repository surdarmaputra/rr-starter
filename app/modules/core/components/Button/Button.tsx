import { Slot } from '@radix-ui/react-slot';
import { cva, type VariantProps } from 'class-variance-authority';
import * as React from 'react';

import { cn } from '~/modules/core/libs/shadcn/utils';

import type { TestableComponentProps } from '../../types';

const buttonVariants = cva(
  `
    inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-semibold
    ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2
    focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none
    disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0
    dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300
  `,
  {
    variants: {
      variant: {
        primary: 'bg-primary text-white hover:bg-primary-hover',
        'primary-outlined':
          'border border-primary text-primary hover:bg-primary-hover hover:text-white',
        'primary-ghost': 'text-primary hover:bg-primary-hover hover:text-white',
        secondary: 'bg-secondary text-white hover:bg-secondary-hover',
        'secondary-outlined':
          'border border-secondary text-secondary hover:bg-secondary-hover hover:text-white',
        'secondary-ghost':
          'text-secondary hover:bg-secondary-hover hover:text-white',
        success: 'bg-success text-white hover:bg-success-hover',
        'success-outlined':
          'border border-success text-success hover:bg-success-hover hover:text-white',
        'success-ghost': 'text-success hover:bg-success-hover hover:text-white',
        warning: 'bg-warning text-white hover:bg-warning-hover',
        'warning-outlined':
          'border border-warning text-warning hover:bg-warning-hover hover:text-white',
        'warning-ghost': 'text-warning hover:bg-warning-hover hover:text-white',
        danger: 'bg-danger text-white hover:bg-danger-hover',
        'danger-outlined':
          'border border-danger text-danger hover:bg-danger-hover hover:text-white',
        'danger-ghost': 'text-danger hover:bg-danger-hover hover:text-white',
        link: 'text-textTitle underline-offset-4 hover:underline',
      },
      size: {
        default: 'h-10 px-4 py-2',
        sm: 'h-9 px-3',
        lg: 'h-11 px-8',
        icon: 'h-10 w-10',
      },
    },
    defaultVariants: {
      variant: 'primary',
      size: 'default',
    },
  },
);

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement>,
    VariantProps<typeof buttonVariants>,
    TestableComponentProps {
  asChild?: boolean;
}

const Button = React.forwardRef<HTMLButtonElement, ButtonProps>(
  ({ className, variant, size, asChild = false, ...props }, ref) => {
    const Comp = asChild ? Slot : 'button';
    return (
      <Comp
        className={cn(buttonVariants({ variant, size, className }))}
        ref={ref}
        {...props}
      />
    );
  },
);

Button.displayName = 'Button';

export { Button, buttonVariants };
