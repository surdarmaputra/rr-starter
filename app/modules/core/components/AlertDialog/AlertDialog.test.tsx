import { fireEvent, render, screen, waitFor, within } from '@testing-library/react';
import { useState } from 'react';
import { describe, expect, test } from 'vitest';

import { Button } from '../Button/Button';
import { AlertDialog, AlertDialogAction, AlertDialogCancel, AlertDialogContent, type AlertDialogContentProps, AlertDialogDescription, type AlertDialogDescriptionProps, AlertDialogFooter, type AlertDialogFooterProps, AlertDialogHeader, AlertDialogTitle, type AlertDialogTitleProps } from './AlertDialog';

interface TestComponentProps {
  contentProps?: AlertDialogContentProps;
  titleProps?: AlertDialogTitleProps;
  descriptionProps?: AlertDialogDescriptionProps;
  footerProps?: AlertDialogFooterProps;
}

export function TestComponent({
  contentProps,
  titleProps,
  descriptionProps,
  footerProps,
}: TestComponentProps) {
  const [isDialogOpen, setIsDialogOpen] = useState<boolean>(false);
  return (
    <>
      <Button onClick={() => setIsDialogOpen(true)}>Toggle Dialog</Button>
      <AlertDialog onOpenChange={setIsDialogOpen} open={isDialogOpen}>
        <AlertDialogContent data-testid='dialog-content-testid' {...contentProps}>
          <AlertDialogHeader >
            <AlertDialogTitle data-testid="dialog-title-testid" {...titleProps}>Are you absolutely sure?</AlertDialogTitle>
            <AlertDialogDescription data-testid='dialog-description-testid' {...descriptionProps}>
              This is a description.
            </AlertDialogDescription>
          </AlertDialogHeader>
          <AlertDialogFooter data-testid='dialog-footer-testid' {...footerProps}>
            <AlertDialogCancel>Cancel</AlertDialogCancel>
            <AlertDialogAction>Continue</AlertDialogAction>
          </AlertDialogFooter>
        </AlertDialogContent>
      </AlertDialog>
    </>
  );
}

describe('AlertDialog', () => {
  function renderComponent({
    contentProps,
    titleProps,
    descriptionProps,
    footerProps,
  }: TestComponentProps = {}) {
    return render(<TestComponent
      contentProps={contentProps}
      descriptionProps={descriptionProps}
      footerProps={footerProps}
      titleProps={titleProps}
    />);
  }

  test('renders with data-testid', async () => {
    renderComponent();
    expect(screen.queryByTestId('dialog-content-testid')).toBeNull();
    fireEvent.click(await screen.findByText(/Toggle Dialog/));

    const dialog = await screen.findByTestId('dialog-content-testid');
    const title = await within(dialog).findByTestId('dialog-title-testid');
    const description = await within(dialog).findByTestId('dialog-description-testid');
    const footer = await within(dialog).findByTestId('dialog-footer-testid');

    await within(title).findByText(/Are you absolutely sure?/);
    await within(description).findByText(/This is a description./);
    await within(footer).findByText(/Cancel/);
    await within(footer).findByText(/Continue/);
  });

  test('renders snapshot', async () => {
    renderComponent();
    fireEvent.click(await screen.findByText(/Toggle Dialog/));
    const dialog = await screen.findByTestId('dialog-content-testid');
    expect(dialog).toMatchSnapshot();
  });

  test('cancels dialog', async () => {
    renderComponent();

    fireEvent.click(await screen.findByText(/Toggle Dialog/));
    const dialog = await screen.findByTestId('dialog-content-testid');

    fireEvent.click(await within(dialog).findByText(/Cancel/));
    await waitFor(() => expect(screen.queryByTestId('dialog-content-testid')).toBeNull());
  });
});
