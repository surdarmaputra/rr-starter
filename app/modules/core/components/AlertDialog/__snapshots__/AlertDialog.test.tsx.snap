// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AlertDialog > renders snapshot 1`] = `
<div
  aria-describedby="radix-:r5:"
  aria-labelledby="radix-:r4:"
  class="fixed left-[50%] top-[50%] z-50 grid w-full max-w-lg translate-x-[-50%] translate-y-[-50%] gap-4 border border-slate-200 bg-white p-6 shadow-lg duration-200 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] dark:border-slate-800 dark:bg-slate-950 sm:rounded-lg"
  data-state="open"
  data-testid="dialog-content-testid"
  id="radix-:r3:"
  role="alertdialog"
  style="pointer-events: auto;"
  tabindex="-1"
>
  <div
    class="flex flex-col space-y-2 text-center sm:text-left"
  >
    <h2
      class="text-lg font-semibold"
      data-testid="dialog-title-testid"
      id="radix-:r4:"
    >
      Are you absolutely sure?
    </h2>
    <p
      class="text-sm text-slate-500 dark:text-slate-400"
      data-testid="dialog-description-testid"
      id="radix-:r5:"
    >
      This is a description.
    </p>
  </div>
  <div
    class="flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
    data-testid="dialog-footer-testid"
  >
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 border border-secondary text-secondary hover:bg-secondaryHover hover:text-white h-10 px-4 py-2 mt-2 sm:mt-0"
      type="button"
    >
      Cancel
    </button>
    <button
      class="inline-flex items-center justify-center gap-2 whitespace-nowrap rounded-full text-sm font-medium ring-offset-white transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-slate-950 focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 dark:ring-offset-slate-950 dark:focus-visible:ring-slate-300 bg-secondary text-white hover:bg-secondaryHover h-10 px-4 py-2"
      type="button"
    >
      Continue
    </button>
  </div>
</div>
`;
